src/main/java/build/buildfarm/worker/ReportResultStage.java

```diff
@@ -186,19 +185,112 @@ protected OperationContext tick(OperationContext operationContext) throws Interr
           .setIsExecutable(Files.isExecutable(outputPath));
       inlineContentBytes = inlineOrDigest(
           content,
-          policy,
+          workerContext.getFileCasPolicy(),
           contents,
           inlineContentBytes,
           workerContext.getInlineContentLimit(),
           () -> outputFileBuilder.setContent(content),
           (fileContent) -> outputFileBuilder.setDigest(getDigestUtil().compute(fileContent)));
     }
 
+    for (String outputDir : outputDirs) {
```

Comment:
I would highly recommend to use an implementation similar to https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/remote/AbstractRemoteActionCache.java#L289
It is much easier to understand (to me, at least), and also more efficient in not computing the digests to upload twice (I guess you can also fix the inefficiency here by maintaining a similar map of digest to path or byte[]).

This will also allow you to optimize later by not loading all the contents to upload into memory at once (I suspect that you will have to do this soon).

Perhaps more importantly, the dirents of each directory need to be sorted per https://github.com/googleapis/googleapis/blob/master/google/devtools/remoteexecution/v1test/remote_execution.proto#L457. IIUC, walkFileTree gives no order guarantees.

Reply:
Having looked at the bazel implementation, I would say it falls short of my ideal. It should, for each path to upload: start a digesting chunker (that computes a digest in a single pass of the file) and on completion decrements a latch count, which releases the action result `build()` and post. I will try to cruft that up.